File: Plugins/uPacketFragmentation/uPacketFragmentation/Fragmenter.cpp
#include <chrono>
#include <algorithm>
#include "Fragmenter.h"
#include "Common.h"
#include "Log.h"


namespace uPacketFragmentation
{


void PacketFragmenter::FragmentData(const void *pData, uint32_t size)
{
    using namespace std::chrono;

    fragments_.clear();

    constexpr uint32_t headerSize = sizeof(PacketHeader);
    const uint32_t maxBufSize = maxPacketSize_ - headerSize;
    const uint32_t fragmentCount = size / maxBufSize + 1;
    const uint32_t fragmentSize = size / fragmentCount + 1;

    PacketHeader header = { GetPacketHeaderVersion() };
    header.type = static_cast<uint32_t>(PacketType::Fragment);
    header.totalSize = size;
    header.timestamp = time_point_cast<milliseconds>(system_clock::now()).time_since_epoch().count();
    header.frameIndex = frameIndex_++;
    header.fragmentCount = fragmentCount;

    for (uint32_t i = 0; i < fragmentCount; ++i)
    {
        const auto bufSize = static_cast<size_t>(maxBufSize) + headerSize;
        auto buf = std::make_unique<char[]>(bufSize);

        header.fragmentIndex = i;
        header.offsetSize = i * fragmentSize;
        header.fragmentSize = std::min(fragmentSize, size - header.offsetSize);
        memcpy_s(buf.get(), headerSize, &header, headerSize);

        const auto *pPartialData = static_cast<const char*>(pData) + header.offsetSize;
        memcpy_s(buf.get() + headerSize, maxBufSize, pPartialData, header.fragmentSize);

        Fragment fragment;
        fragment.size = header.fragmentSize + headerSize;
        fragment.buf = std::move(buf);
        fragments_.push_back(std::move(fragment));
    }
}


uint32_t PacketFragmenter::GetFragmentSize(uint32_t index) const
{
    if (index >= GetFragmentCount()) return 0;

    const auto &fragment = fragments_.at(index);
    return fragment.size;
}


const char * PacketFragmenter::GetFragmentData(uint32_t index) const
{
    if (index >= GetFragmentCount()) return nullptr;

    const auto &fragment = fragments_.at(index);
    return fragment.buf.get();
}


}

File: Assets/uPacketFragmentation/Scripts/Fragmenter.cs
namespace uPacketFragmentation
{

public class Fragmenter : System.IDisposable
{
    public uint id { get; private set; }

    public Fragmenter()
    {
        id = Lib.Fragmenter.Create();
    }

    ~Fragmenter()
    {
        Dispose();
    }

    public void Dispose()
    {
        Lib.Fragmenter.Destroy(id);
    }

    public bool isValid
    {
        get { return Lib.Fragmenter.IsValid(id); }
    }

    public int maxPacketSize
    {
        set
        {
            Lib.Fragmenter.SetMaxPacketSize(id, (uint)value);
        }
    }

    public void Fragment(System.IntPtr data, uint size)
    {
        Lib.Fragmenter.FragmentData(id, data, size);
    }

    public uint GetFragmentCount()
    {
        return Lib.Fragmenter.GetFragmentCount(id);
    }

    public System.IntPtr GetFragmentData(uint index)
    {
        return Lib.Fragmenter.GetFragmentData(id, index);
    }

    public uint GetFragmentSize(uint index)
    {
        return Lib.Fragmenter.GetFragmentSize(id, index);
    }
}

}
